### PR TITLE
chore: adopt turborepo (#5316

### PR DESCRIPTION
**Pull request by ovflowd**
This PR does a massive infrastructure change by doing the following:

- Adopting `turborepo` for blazing fast builds and pipelines
- Updated docs with the usage of TurboRepo
- Removed support to IE10/IE11
- Removed `npm-run-all`
- Added missing dependencies
- Update `package.json` scripts

### CODE REVIEW

**Comment by ovflowd** [comment]
Hey, @nodejs/website, this is a massive change in the underlying infrastructure of this repository. Hence this PR cannot be fast-tracked. Please, everyone give a review.

File: .github/workflows/github-pages.yml [hunk 1]
```diff
@@ -24,27 +24,33 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v3
 
+      - name: Install PNPM
+        uses: pnpm/action-setup@v2
+        with:
+          version: latest
+
```
Comment (aduh95) [inline .github/workflows/github-pages.yml]:
```suggestion
```

File: .github/workflows/github-pages.yml [hunk 2]
```diff
@@ -24,27 +24,33 @@ jobs:
...
       - name: Set up Node.js
         uses: actions/setup-node@v3
         with:
           node-version-file: '.nvmrc'
-          cache: npm
+          cache: pnpm
 
       - name: Install NPM packages
-        run: npm ci
+        run: pnpm install --frozen-lockfile
```
Comment (aduh95) [inline .github/workflows/github-pages.yml]:
```suggestion
        run: corepack pnpm install --frozen-lockfile
```

File: .github/workflows/github-pages.yml [hunk 3]
```diff
@@ -54,10 +60,11 @@ jobs:
         with:
           path: |
             .next/cache
-          key: ${{ runner.os }}-nextjs-${{ hashFiles('**/package-lock.json') }}-${{ hashFiles('.next/cache/eslint') }}
+            node_modules/.cache
+          key: ${{ runner.os }}-nextjs-${{ hashFiles('**/pnpm-lock.yaml') }}-${{ hashFiles('node_modules/.cache') }}
 
       - name: Export Next.js static files
-        run: npm run export
+        run: pnpm export
```
Comment (aduh95) [inline .github/workflows/github-pages.yml]:
```suggestion
        run: corepack pnpm export
```

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -16,17 +13,36 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v3
 
+      - name: Install PNPM
+        uses: pnpm/action-setup@v2
+        with:
+          version: latest
+
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
```

File: .github/workflows/pull-request.yml [hunk 5]
```diff
@@ -16,17 +13,36 @@ jobs:
...
       - name: Set up Node.js
         uses: actions/setup-node@v3
         with:
           node-version-file: '.nvmrc'
-          cache: npm
+          cache: pnpm
 
       - name: Install NPM packages
-        run: npm ci
+        run: pnpm install --frozen-lockfile
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
        run: corepack pnpm install --frozen-lockfile
```

File: .github/workflows/pull-request.yml [hunk 6]
```diff
@@ -41,17 +57,36 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v3
 
+      - name: Install PNPM
+        uses: pnpm/action-setup@v2
+        with:
+          version: latest
+
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
```

File: .github/workflows/pull-request.yml [hunk 7]
```diff
@@ -41,17 +57,36 @@ jobs:
...
       - name: Set up Node.js
         uses: actions/setup-node@v3
         with:
           node-version-file: '.nvmrc'
-          cache: npm
+          cache: pnpm
 
       - name: Install NPM packages
-        run: npm ci
+        run: pnpm install --frozen-lockfile
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
        run: corepack pnpm install --frozen-lockfile
```

File: .github/workflows/pull-request.yml [hunk 8]
```diff
@@ -66,30 +101,37 @@ jobs:
       - name: Git Checkout
         uses: actions/checkout@v3
 
+      - name: Install PNPM
+        uses: pnpm/action-setup@v2
+        with:
+          version: latest
+
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
```

File: .github/workflows/pull-request.yml [hunk 9]
```diff
@@ -66,30 +101,37 @@ jobs:
...
       - name: Set up Node.js
         uses: actions/setup-node@v3
         with:
           node-version-file: '.nvmrc'
-          cache: npm
+          cache: pnpm
 
       - name: Install NPM packages
-        run: npm ci
+        run: pnpm install --frozen-lockfile
```
Comment (aduh95) [inline .github/workflows/pull-request.yml]:
```suggestion
        run: corepack pnpm install --frozen-lockfile
```

**Review by aduh95 — COMMENTED** [review]
I suggest we use Corepack in CI.

File: .github/workflows/github-pages.yml (hunk 1 above)
Comment (ovflowd) [reply .github/workflows/github-pages.yml]: Hey @aduh95 any instructions for using Corepack?

**Comment by ovflowd** [comment]
@aduh95 quick question: Is Corepack bundled already with `actions/setup-node`?

I guess my question is if I need to enable corepack or do something else.

File: .github/workflows/github-pages.yml (hunk 1 above)
Comment (ovflowd) [reply .github/workflows/github-pages.yml]: Because if I don't provide the install action for PNPM, then `actions/setup-node@v3` fails

**Comment by ljharb** [comment]
why would node, which ships npm, want to adopt "not npm" as its package manager?

**Comment by ovflowd** [comment]
> why would node, which ships npm, want to adopt "not npm" as its package manager?

That is a fair argument. I'm pretty much just experimenting. I wanted to see the differences of installation speeds from `pnpm` to `npm`.

But after some tests, it seems to be just around 4-6 seconds, which is not an improvement. I'm going to revert to `npm`.

Note.: This was an Experiment, and I wanted to hear what other Core Collaborators and TSC would have to say about such change.

File: .github/workflows/github-pages.yml (hunk 1 above)
Comment (aduh95) [reply .github/workflows/github-pages.yml]: AFAIK there's shouldn't need any install step as it is installed by default in GHA runner along `node`.

**Comment by AugustinMauroy** [comment]
Personally -1 for `pnpm`
But +1 for `turborepo` and +1 to remove IE 11

**Comment by aduh95** [comment]
> why would node, which ships npm, want to adopt "not npm" as its package manager?

Node.js shipping is not an endorsement, it's based on historical reasons. And if this repo ends up adopting a solution that's not npm to manage its dependencies, it would be very wrong to say "node adopts “not npm” as its package manager", this is not node, it's the nodejs.org repo.


> I wanted to hear what other Core Collaborators and TSC would have to say about such change.

I'd be fine with it, the maintainers of this repo should have total freedom on this kind of technical choice because it shouldn't matter to anyone but them :)

**Comment by ovflowd** [comment]
cc @nodejs/website can we get reviews here? cc @nodejs/build (if y'all have anything against turborepo)

**Comment by ovflowd** [comment]
@ aduh95 arguments are fair. I will, for now, not switch to `pnpm` as I want to avoid possible backfire or discontent from other Collaborators as this seems a mixed feeling.

But I'd revisit this in the future, given the matter.

Since this PR now only introduces an adoption to Turborepo, I'm proceeding with merging this as it is. cc @nodejs/website @nodejs/build for visibility.

**Comment by targos** [comment]
Do we need to change how the site is built on the production server or it will still work fine without `npx turbo` ?

**Comment by ovflowd** [comment]
> Do we need to change how the site is built on the production server or it will still work fine without npx turbo ?

Works completely fine without `turbo`. The `turbo` command is for local development (optional) and Vercel Servers. But it is entirely optional.

**Pull request merged**